### ISSUE TITLE
Documentation: Use of (JSON) Maps + Type Annotation Parentheses

### ISSUE THREAD

**Issue by gatsbysghost** [dhall-lang/dhall-lang#1292]
Hey there! I recently implemented Dhall in a critical capacity for my job, and one of the simplest things that took me the longest to figure out was the proper use of the `Map` type. This Issue is filed to call out an opportunity to improve the clarity of documentation (and possibly error reporting, to the extent that such a thing is possible and desirable).

# Scenario That Precipitated This Whole Thing
## Context
The way that I'm deploying Dhall at work is basically super simple: we maintain different regional "clusters" of servers, and some server/application configuration items differ depending on the cluster. So the basic structure of the thing looks like this:

```haskell
let cluster_config
      : Type
      = { cluster_tld : Text
        , example_field : Optional Text
        }

let primary
     : cluster_config
     = { cluster_config = "primary.example.com"
       , example_field = Some "Example value"
       }

in { clusters = { primary } }
```

## The Part Where I Got Super Confused
So far, so good--it's a really really basic use case. But I was asked by one of my coworkers to incorporate something that would end up looking like this in the result JSON:

```json
"gateways": {
    "NA": ["1.2.3.4", "1.2.3.5"],
    "APAC": ["11.12.13.14", "11.12.13.15"]
}
```

...so my instinct was to basically find a way to work this kind of structure in:
```haskell
let primary
     : cluster_config
     = { cluster_config = "primary.example.com"
       , example_field = Some "Example value"
       , gateways = { NA = ["1.2.3.4", "1.2.3.5"]
                    , APAC = ["11.12.13.14", "11.12.13.15"]
                    }
       }
```

But the trouble with this is that it's not clear to me what kind of type the `gateways` field would be--it seems like it should maybe be a bare `Record`, possibly something like `{ NA : List Text, APAC : List Text }`. **But what if the keys are not possible to predict in advance?** (i.e., what if `NA` and `APAC` are not the only options, and we cannot predict at compile-time what the keys in this map will be?)

I assumed based on the documentation that Map would be the right way to represent this data, but the documentation on `Map` is really heavy on type theory and set theory in a way that makes it kind of difficult to interpret for a workaday SRE like me who does not come from a strong type-theoretical or FP/Haskell background. So I found the following things really unclear as they pertained to solving for the (fairly common, I think?) question of how to represent the type of a Map (from text to a list of text).

### The First Thing I Tried
So--again, looking for a `Map` of `Text` -> `List Text`. My first guess for this was something like:

```haskell
let Map = https://prelude.dhall-lang.org/v22.0.0/Map/Type sha256:210c7a9eba71efbb0f7a66b3dcf8b9d3976ffc2bc0e907aadfb6aa29c333e8ed

let cluster_config
      : Type
      = { cluster_tld : Text
        , example_field : Optional Text
        , gateways : Map Text List Text
        }

let primary
     : cluster_config
     = { cluster_config = "primary.example.com"
       , example_field = Some "Example value"
       , gateways = toMap( { NA = ["1.2.3.4", "1.2.3.5"]
                           , APAC = ["11.12.13.14", "11.12.13.15"]
                           }
                    )
       }

in { clusters = { primary } }
```

...but this turns up the following error from `dhall-to-json`:
```
Map : Type → Type → Type

Error: Wrong type of function argument

- Type
+ … → … (a function type)

[...]

You tried to invoke the following function:

↳ Map Text

... which expects an argument of type or kind:

↳ Type

... on the following argument:

↳ List

... which has a different type or kind:

↳ Type → Type
```

Now--given that I really thought I **did** feed Map three types, that led me down the path of a bunch of wrong ideas:

### The Part Where I Tried A Bunch Of Wrong Ideas That Are Only Obviously Incorrect In Retrospect
So my first instinct was, looking at the error output: Okay, let's go look at [the documentation for `Map`](https://docs.dhall-lang.org/tutorials/Language-Tour.html#maps), and lo, an explanation:

![Screen Shot 2022-08-02 at 8 47 18 PM](https://user-images.githubusercontent.com/4369667/182519874-4c61a268-98ac-40de-b8e2-817cab8e48c4.png)

So on that basis I tried a bunch of really wrong ideas that I won't bore you with here; my first thought was that I was getting too clever by half with the type annotation; I reasoned that because the docs say that a `Map` is underlyingly a `List`, I should maybe try to reference it that way (i.e., that the actual issue was that `Map` isn't a proper type for the purposes of annotation). So I tried things like `List Text List Text` (haha, I know--everybody laugh at the silly non-CS-major; it is indeed very funny, and it's funnier still that I might have accidentally fixed it with an annotation of `List { mapKey : Text, mapValue : List Text }`, but I honestly just didn't understand the `Map` type).

## The Correct Thing, Incidentally
I'm sure you've already spotted what the right answer is, but for the record:

```haskell
let Map = https://prelude.dhall-lang.org/v22.0.0/Map/Type sha256:210c7a9eba71efbb0f7a66b3dcf8b9d3976ffc2bc0e907aadfb6aa29c333e8ed

let cluster_config
      : Type
      = { cluster_tld : Text
        , example_field : Optional Text
        , gateways : Map Text (List Text)
        }

let primary
     : cluster_config
     = { cluster_config = "primary.example.com"
       , example_field = Some "Example value"
       , gateways = toMap( { NA = ["1.2.3.4", "1.2.3.5"]
                           , APAC = ["11.12.13.14", "11.12.13.15"]
                           }
                    )
       }

in { clusters = { primary } }
```
The thing that sent me down the wrong (but ultimately quite educational) path was just that I needed some parentheses! Jeepers creepers; I have been working in software for 7 years.

# I Have Notes
So, my main thoughts here are as follows:
1. The `--explain` for a type mismatch of `Type -> Type -> Type` with a `Map` might do well to suggest the use of parentheses?
2. The docs might benefit from an example showing how to construct this type of JSON data structure (a `Map` from an arbitrary number of non-regularized strings to a list of strings). I would also have benefited from e.g. a full, basic `Map Text Text` example. Maps are a super common use case for configuration, and I feel like the docs for the `Map` type and `toMap` function are generally the hardest to pick up and go from as a developer guide; they are also not really discussed in `Learn Dhall in Y Minutes`, either--which, in case you're not interested in modifying the docs, I would be happy to submit an update to.
3. Finally, I think the docs might benefit from touching on the use cases for parentheses in Dhall type annotations; this was a really easy thing to get tripped up on, and I don't remember reading anything about parentheses in type annotations, honestly.

Thank you so much for everything you've contributed to Dhall! I am a huge fan of your excellent work and I thank you for your consideration! 😄

**Comment by Gabriella439** [comment]
So my intuition here is to lean most heavily on improving the type error since it's going to be hit-or-miss for the documentation to anticipate every possible way that things go wrong.

In this case, though, the type error is pretty direct about what went wrong:

```
Map : Type → Type → Type

Error: Wrong type of function argument

- Type
+ … → … (a function type)

[...]

You tried to invoke the following function:

↳ Map Text

... which expects an argument of type or kind:

↳ Type

... on the following argument:

↳ List

... which has a different type or kind:

↳ Type → Type
```

I think the only way it could have been more direct is to underline things a bit?

**Comment by gatsbysghost** [comment]
Yeah, and truly it's super legible if your head is already in Dhall or Haskell or FP, but I think my thing is that the documentation or the `explain` may miss a bit for people who only have an imperative vocabulary, or who are new to the language.

So, a little more concretely, if I put my imperative/OOP-pilled (or new Dhaller) hat on, I might see:

```
Map : Type → Type → Type

Error: Wrong type of function argument

- Type
+ … → … (a function type)
```

...and (justifiably, I think?) say to myself, "Well, I did give you 4 `Type` arguments for Map, and I get that that's too many, but...does that mean that I need to give you a bare `List` without a Type annotation? [tries that.] No, that doesn't work--so does it mean that you simply can't have a `List` in this position because that causes it to be too many arguments? Or what?"

So part of me thinks that when a user provides a `Type -> Type` (or even specifically a `List`) where there should be a `Type`, it might be useful to hint to them that they can use `(Type -> Type)`. (I feel like this is the kind of constructive error that `rustc` tends to give me :) ). Unless you can think of a use case where that might be bad advice (or if it's prohibitively hard to implement).

I think I'd also be kind of baffled by:

```
- Type
+ … → … (a function type)
```

Looking at this as a new user & forcibly removing all the context from my brain, my questions are:
- Is the `-` value the type it *should* be or the type we *found*?
- What does any of the line "... -> ... (a function type)" mean?
- What has been elided out with the `...`s?
- What part or parts of this does `(a function type)` refer to? Does this just mean `+ Type -> Type`?

I look at this and I guess I don't really see the benefit of going with `+`/`-` and using obfuscating `...`s while we have the type data and the full expression; I guess I wonder what downsides would be exposed by giving the user that info.

So putting it all together, I think the error output I would like to see would be a little more like this:

```
Expected: Map Type -> Type -> Type

Found:    Map Type -> Type -> Type -> Type
                              ^^^^^^^^^^^^

If you are trying to use a `Type -> Type` (e.g. a List) where a single `Type` is required, you can
use parentheses to consolidate Types. For example:

List Text
^^^^^^^^^
has the type annotation: Type -> Type
                         ^^^^^^^^^^^^
...but (List Text)
       ^^^^^^^^^^^
has the type annotation: Type
                         ^^^^
```

I think as it goes, I might also want to propose a change to `Learn Dhall in Y Minutes` or the language tour to include parentheses in an example, just to show their syntactic utility here. Parentheses were a shot-in-the-dark guess for me when I tried them, and nowhere, as far as I can tell, do the docs tell me that I can consolidate `Types` this way--though I guess it's common knowledge for Haskell people? 😅
